### PR TITLE
RTL6c: Publish while not connected

### PR DESCRIPTION
**Pull request by tcard**
- [x] [RTL6c1](https://docs.ably.io/client-lib-development-guide/features/#RTL6c1)
- [x] [RTL6c2](https://docs.ably.io/client-lib-development-guide/features/#RTL6c2)
- [x] [RTL6c4](https://docs.ably.io/client-lib-development-guide/features/#RTL6c4)
- [x] [RTL6c5](https://docs.ably.io/client-lib-development-guide/features/#RTL6c5)

This PR also includes random fixes that came up during testing, including the culprit of #172's CI failure.

### CODE REVIEW

File: ably/realtime_channel.go [hunk 1]
```diff
@@ -462,24 +428,65 @@ func (c *RealtimeChannel) History(params *PaginateParams) (*PaginatedResult, err
...
 
 func (c *RealtimeChannel) send(msg *proto.ProtocolMessage) (Result, error) {
-	if _, err := c.attach(false); err != nil {
-		return nil, err
-	}
-	res, listen := newErrResult()
-	switch c.State() {
-	case ChannelStateInitialized, ChannelStateAttaching:
-		c.queue.Enqueue(msg, listen)
+	if res, enqueued := c.maybeEnqueue(msg); enqueued {
```
Comment (lmars) [inline ably/realtime_channel.go:431]: Given `Publish` is now blocking but takes a context, should this queue not take into account the context deadline? For example, if a connection is stuck in a connecting / disconnecting loop and I publish with a 5s timeout, I'd probably expect that to return with an error after 5s?

Comment (tcard) [reply ably/realtime_channel.go:431]: The contexts (here and elsewhere) are handled at the outer layer. The pattern is:

* Public function takes a context and calls an internal function.
* Internal function starts some process and returns a Result to wait for it. (Here, the Result waits for the message to be dequeued, and either be failed or published and then acknowledged.)
* Public function waits for the Result, but will stop waiting if the context is called.

This implements [what we decided](https://github.com/ably/ably-go/issues/197#issuecomment-696979466) in terms of what a context means in the Realtime API. (In REST, contexts will be passed all the way to the HTTP library.)

File: ably/realtime_channel.go [hunk 2]
```diff
@@ -462,24 +428,65 @@ func (c *RealtimeChannel) History(params *PaginateParams) (*PaginatedResult, err
...
-	default:
-		return nil, &ErrorInfo{Code: 90001}
 	}
+
+	if !c.canSend() {
+		return nil, newError(ErrChannelOperationFailedInvalidChannelState, nil)
+	}
+
+	res, listen := newErrResult()
 	if err := c.client.Connection.send(msg, listen); err != nil {
```
[the remark this answers is not in the document]
Comment (tcard) [reply ably/realtime_channel.go:440]: That's what [RTN19](https://docs.ably.io/client-lib-development-guide/features/#RTN19) is about (being implemented at #170).

File: ably/realtime_channel.go (hunk 1 above)
Comment (lmars) [reply ably/realtime_channel.go:431]: Ah I'm sorry, I thought that `Publish` would not return when the context was cancelled, but I see I was incorrect 👍

File: ably/realtime_channel.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lmars) [reply ably/realtime_channel.go:440]: Ah ok thanks

**Review by lmars — APPROVED** [review]

**Review by gernest — APPROVED** [review]

**Pull request merged**